f/connectors/kobotoolbox/kobotoolbox_responses.py

```diff
@@ -458,17 +459,51 @@ def _safe_insert(cursor, table_name, columns, values):
 
         Returns
         -------
-        None
+        tuple
+            A tuple containing the count of inserted and updated rows.
         """
+        inserted_count = 0
+        updated_count = 0
+
+        # Check if there is an existing row that is different from the new values
+        id_index = columns.index("_id")
+        values[id_index] = str(values[id_index])
+        select_query = sql.SQL("SELECT {fields} FROM {table} WHERE _id = %s").format(
+            fields=sql.SQL(", ").join(map(sql.Identifier, columns)),
+            table=sql.Identifier(table_name),
+        )
+        cursor.execute(select_query, (values[columns.index("_id")],))
+        existing_row = cursor.fetchone()
+
+        if existing_row and list(existing_row) == values:
+            # No changes, skip the update
+            return inserted_count, updated_count
```

Comment:
Why do we need everything above?  Won't the INSERT below just re-write the same values and the end result is the same?

Reply:
ah, is it so you can more accurately report on inserted_count, updated_count, without considering an exact overwrite to be an update?

Reply:
Yep, that's correct - otherwise everything that isn't inserted is counted as updated.

We absolutely do not _need_ it for the core ETL functionality of the script, but arguably, it's a useful feature for Windmill operators to understand what their runs are doing each time. 

It could be interesting to benchmark how the script performs with and without this code for the largest Kobo dataset that we are ingesting.

Reply:
I would document this, too, then.

Otherwise looks good to me!